### PR TITLE
Debug Mode c++ macro

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Implements a minimal C++ macro that can support the base case of features described here: #99918

And adds to embedding table.

### Perf:
Ran a script to see if adding this path impacts perf: https://gist.github.com/drisspg/d9abae0f7f40f7b924edf9cffff2b12e
Seems like it is below noise levels on my machine

### CODE REVIEW

File: test/test_debug_mode.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Owner(s): ["module: nn"]
+
+import contextlib
+import os
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (janeyx99) [inline test/test_debug_mode.py:8]: unused

File: test/test_debug_mode.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from torch.testing._internal.common_device_type import (
+    instantiate_device_type_tests,
+)
+from torch.testing._internal.common_nn import NNTestCase
+from torch.testing._internal.common_utils import NOTEST_CPU, run_tests
```
Comment (janeyx99) [inline test/test_debug_mode.py:13]: Could import TestCase here and use that instead of NNTestCase

File: c10/util/Exception.h [hunk 3]
```diff
@@ -534,6 +535,17 @@ namespace detail {
         static_cast<uint32_t>(__LINE__),           \
         TORCH_CHECK_MSG(cond, "", ##__VA_ARGS__)); \
   }
+// This Macro is used to check conditions that are potentially expensive to
+// check and are only checked in debug mode. It is not recommended to use this
+// macro to check conditions that are cheap to check.
+#define TORCH_DEBUG_CHECK(cond, ...)                                 \
+  if (std::getenv("TORCH_DEBUG") && C10_UNLIKELY_OR_CONST(!(cond))) { \
```
Comment (cpuhrsch) [inline c10/util/Exception.h:542]: smth smth performance

File: c10/util/Exception.cpp [hunk 4]
```diff
@@ -114,6 +114,13 @@ void torchInternalAssertFail(
   torchCheckFail(func, file, line, c10::str(condMsg, userMsg));
 }
 
+bool env_flag_set(const char* env_var_name) {
+  const char* const env_string = std::getenv(env_var_name);
+  return (env_string == nullptr) ? false : std::strcmp(env_string, "0");
+}
+
+C10_EXPORT bool TORCH_RUNTIME_DEBUG = env_flag_set("TORCH_RUNTIME_DEBUG");
```
Comment (drisspg) [inline c10/util/Exception.cpp:122]: So I haven't seen this pattern anywhere and I feel like there is probably a reason for this. 

I verified with lldb that this is hit once during loading of libc10
``` Shell
frame #0: 0x00007ffff70e0d64 libc10.so`::__cxx_global_var_init.9() at Exception.cpp:122:39
```
std::getenv is labeled no_execpt() but still idk feels a little sketch. I think maybe I export the symbol but just set it to false, and do an initial os.getenv() in torch/__init__.cpp ?

File: aten/src/ATen/native/transformers/attention.cpp [hunk 5]
```diff
@@ -495,6 +496,58 @@ inline void validate_sdpa_input(
   }
   return;
 }
+
+double get_min_value(caffe2::TypeMeta dtype){
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:500]: I wrestled tooooooooooooooo long trying to find the pattern for this. There does exists a ScaleTypetoCPPType I thought I had something but was having trouble

**Comment by drisspg** [comment]
Good use case https://github.com/pytorch/pytorch/pull/113433

**Comment by drisspg** [comment]
https://www.evanjones.ca/setenv-is-not-thread-safe.html

**Pull request closed without merging**